Add default implementation for tracking metadata

sourabhmarathe opened the pull request:
Adds NetworkTracker trait that allows users to implement interface
between their metadata and their view of NetworkGraph. Users can
implement their own metadata structs, but at a minimum will have to
implement the UpdateMetadata trait (even if the functions do nothing).

By default, this trait uses a new struct called DefaultMetadata
which just tracks previously failed routes, and can find out if a
route has previously failed.

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
+pub trait UpdateMetadata {
+	/// Adds a failed route to track. If we see the route again, we'll reject it or we can take
+	/// some action (like set its fee to make route-finding not find that route again)
+	fn add_failed_route(&mut self, route: Vec<Vec<RouteHop>>);
```
I dont think this needs the double-vec - we only care about one path of an MPP failing, not the full route.

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
+	/// some action (like set its fee to make route-finding not find that route again)
+	fn add_failed_route(&mut self, route: Vec<Vec<RouteHop>>);
+	/// If a route starts working again, then we can take it off the failed route list.
+	fn remove_failed_route(&mut self, route: Vec<Vec<RouteHop>>);
```
Hmm, I don't think this needs to exist because there isn't really a good way to use it - the trait is really the list of things that rust-lightning will call on implementations (but the struct can have its own impl with its own extra methods users may want). There isn't ever a good criteria for when rust-lightning should decide that a path should be retired (and even if there were, it would mean rust-lightning would need to be track the list of previously-failed routes to tell the UpdateMetadata when to allow it again).

What you may want, instead, is a route_succeeded() call which rust-lightning could use to tell the object that a route passed.

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
 }
 
+/// Allows for updating user's network metadata.
+pub trait UpdateMetadata {
```
Maybe a better name is RouteFeePenalty or RouteScorer?

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
+	fn add_failed_route(&mut self, route: Vec<Vec<RouteHop>>);
+	/// If a route starts working again, then we can take it off the failed route list.
+	fn remove_failed_route(&mut self, route: Vec<Vec<RouteHop>>);
+	/// Returns true if a route is on the failed routes list, false otherwise.
```
Same here, I think the only thing we need is get_channel_fee_penalty, which can be set to u64::max to emulate this.

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
+	/// Returns true if a route is on the failed routes list, false otherwise.
+	fn check_route(&self, route: Vec<Vec<RouteHop>>) -> bool;
+	/// Sets the channel fee penalty being tracked in the Metadata object.
+	fn set_channel_fee_penalty(&mut self, chan_id: u64, fee_penalty: u64);
```
Same here - rust-lightning will never call this so it doesnt need to be in the trait.

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
+pub struct DefaultMetadata {
+	/// A lits of failed routes. Referenced by the failed route trackers, and sets channel fee
+	/// penalty to a high number if in trait
+	failed_routes: Vec<Vec<Vec<RouteHop>>>,
```
I think we should just track individual channels as "previously failed" and mark those as "infinite" penalty. Probably removing them from the list after a few calls to a route-succeeded function.

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
 }
 
+/// Allows for updating user's network metadata.
+pub trait UpdateMetadata {
```
Seems to me this trait is trying to encompass both inputs consuming to compute score (`add_failed_route`) and scores consumption (`get_channel_fee_penalty`). Second is likely to be consumed by Router and any-kind of peer selection, but first-one is really tied about the origin of data consumed. 

If it's event-based, like receiving a PaymentSuccess or PaymentFailed, client is going to call a `add_failed_route` method-like. If it's self-running, like a thread probing/tracking peer liveness,, you just call something like `spawn_liveness_thread`. So I would favor dissociate those traits even if their are implemented on the same default struct.

sourabhmarathe replied:
I can disassociate these traits, I think that makes sense. Maybe the NetworkGraph <---> User Metadata connection should be called UpdatesMetadata, and the User metadata <---> router should be called ScoresChannels. The user would implement both traits, and the network would, by default, send certain data to the traits, and the router would by, default, ask for channel scores when conducting its A* search.

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
+}
+
+/// A default metadata object that is used to implement the default functionality for the
+/// NetworkTracker trait. A user could  make their own Metadata object and extend the
```
I think there is a question of code sharing here, if people start to work on fancy scoring logic, they may want to combine them, and having each of them implement a custom UpdateMetadata they won't be able to do so.

TheBlueMatt replied:
I don't see why? You could write a combiner, but ultimately we *have* to have a way to linearize any kind of fancy scoring logic users want into one parameter (namely fee paid) so that we can calculate a route based on it. No matter what users do, they have to be able to linearize it.

ariard replied:
Yes I think we agree on the need to linearize it, my point was more on dissociating linearization method (`get_channel_fee_penalty`) from score-feeding ones (`route_succeeded`/`route_failed`). Linearization one should be on such combiner. I don't think we have to do this now, but we may have to introduce such combiner in the future.

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
+	}
+
+	/// Store routes that have previously failed by adding them to a list being held in Metadata
+	fn track_previously_failed_route(&self, mut network_metadata: T, failed_route: Vec<Vec<RouteHop>>) {
```
Why instead of tracking failed route, we don't score really badly the failing link until further notice ? That way it shouldn't be part of next route attemp unless you exhaust all possibilities. Or is there something I'm missing with our A* router.

sourabhmarathe replied:
Yup, I think this is what I have now. I score a channel u64::MAX, and after consecutively participating in enough successful payments, the channel is removed from the failure list.

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,64 @@ impl Readable for NodeInfo {
+pub trait RouteFeePenalty {
+	/// Gets a channel's fee penalty based on its channel_id (as stored in a NetworkGraph object).
+	fn get_channel_fee_penalty(&self, chan_id: u64) -> Option<&u64>;
+	/// Informs metadata object that a route has successfully executed its payment.
```
Maybe add a route_failed?

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,64 @@ impl Readable for NodeInfo {
+/// Allows for updating user's network metadata.
+pub trait RouteFeePenalty {
+	/// Gets a channel's fee penalty based on its channel_id (as stored in a NetworkGraph object).
+	fn get_channel_fee_penalty(&self, chan_id: u64) -> Option<&u64>;
```
I dont know that this needs the `Option<&` part - its easy to just return 0 if there is no penalty.

sourabhmarathe commented:
I added the route_succeeded method to the RoutingFeePenalty trait and added a parameter to get_route that must implement the RoutingFeePenalty trait. I was also curious to hear what you think of the idea to separate the data feeding (eg methods used to update user metadata) and the data consumption methods (eg getting routing_fee from the user).

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -221,7 +221,7 @@ pub fn get_route<C: Deref, L: Deref>(our_node_id: &PublicKey, net_graph_msg_hand
 		( $chan_id: expr, $src_node_id: expr, $dest_node_id: expr, $directional_info: expr, $chan_features: expr, $starting_fee_msat: expr ) => {
 			//TODO: Explore simply adding fee to hit htlc_minimum_msat
-			if $starting_fee_msat as u64 + final_value_msat >= $directional_info.htlc_minimum_msat {
+			if $starting_fee_msat as u64 + final_value_msat >= $directional_info.htlc_minimum_msat + net_metadata.get_channel_fee_penalty($chan_id.clone()) {
```
This line is checking whether we've reached the HTLC minimum and can use this channel, fee penalty should likely appear two lines down in the new_fee calculation

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,71 @@ impl Readable for NodeInfo {
+			let chan_id = route_hop.short_channel_id;
+			let successes_needed = self.failed_channels.get(&chan_id);
+			if successes_needed == None {
+				self.failed_channels.insert(chan_id, 5);
```
I'm super confused here. Looks like you're inserting channels which will be considered "failed" in get_channel_fee_penalty on the basis that they succeeded?

sourabhmarathe replied:
Yes, that's right. It's definitely a bug, I removed this branch of the if statement

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -162,7 +162,7 @@ struct DummyDirectionalChannelInfo {
 /// *is* checked as they may change based on the receiving node.
 pub fn get_route<C: Deref, L: Deref>(our_node_id: &PublicKey, net_graph_msg_handler: &NetGraphMsgHandler<C, L>, target: &PublicKey, first_hops: Option<&[channelmanager::ChannelDetails]>,
-	last_hops: &[RouteHint], final_value_msat: u64, final_cltv: u32, logger: L) -> Result<Route, LightningError> where C::Target: ChainWatchInterface, L::Target: Logger {
+	last_hops: &[RouteHint], final_value_msat: u64, final_cltv: u32, logger: L, net_metadata: impl RouteFeePenalty) -> Result<Route, LightningError> where C::Target: ChainWatchInterface, L::Target: Logger {
```
You can do, it builds and make trait parameters easier to read IMO:

```
diff --git a/lightning/src/routing/router.rs b/lightning/src/routing/router.rs
index 044f8a9f..e6b8541d 100644
--- a/lightning/src/routing/router.rs
+++ b/lightning/src/routing/router.rs
@@ -161,8 +161,8 @@ struct DummyDirectionalChannelInfo {
 /// The fees on channels from us to next-hops are ignored (as they are assumed to all be
 /// equal), however the enabled/disabled bit on such channels as well as the htlc_minimum_msat
 /// *is* checked as they may change based on the receiving node.
-pub fn get_route<C: Deref, L: Deref>(our_node_id: &PublicKey, net_graph_msg_handler: &NetGraphMsgHandler<C, L>, target: &PublicKey, first_hops: Option<&[channelmanager::ChannelDetails]>,
-       last_hops: &[RouteHint], final_value_msat: u64, final_cltv: u32, logger: L, net_metadata: impl RouteFeePenalty) -> Result<Route, LightningError> where C::Target: ChainWatchInterface, L::Target: Logger {
+pub fn get_route<C: Deref, L: Deref, N: RouteFeePenalty>(our_node_id: &PublicKey, net_graph_msg_handler: &NetGraphMsgHandler<C, L>, target: &PublicKey, first_hops: Option<&[channelmanager::ChannelDetails]>,
+       last_hops: &[RouteHint], final_value_msat: u64, final_cltv: u32, logger: L, net_metadata: N) -> Result<Route, LightningError> where C::Target: ChainWatchInterface, L::Target: Logger {
        // TODO: Obviously *only* using total fee cost sucks. We should consider weighting by
        // uptime/success in using a node in the past.
        if *target == *our_node_id {
```

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -450,17 +448,18 @@ pub struct DefaultMetadata {
-	fn set_channel_fee_penalty(&mut self, chan_id: u64, fee_penalty: u64) { self.channel_fee_penalty.insert(chan_id, fee_penalty); }
+impl RouteFeePenalty for DefaultMetadata {
 	fn get_channel_fee_penalty(&self, chan_id: u64) -> Option<&u64> { self.channel_fee_penalty.get(&chan_id) }
+	fn route_succeeded(&mut self, route: Vec<RouteHop>) {
```
Or without a `retain` method on `BTreeMap`, you can do:

```
diff --git a/lightning/src/routing/network_graph.rs b/lightning/src/routing/network_graph.rs
index f0348313..33f54973 100644
--- a/lightning/src/routing/network_graph.rs
+++ b/lightning/src/routing/network_graph.rs
@@ -460,14 +460,13 @@ impl RouteFeePenalty for DefaultMetadata {
        fn route_succeeded(&mut self, route: Vec<RouteHop>) {
                for route_hop in route {
                        let chan_id = route_hop.short_channel_id;
-                       let successes_needed = self.failed_channels.get(&chan_id);
-                       if successes_needed != None {
-                               let dec_successes_needed = successes_needed.unwrap() - 1;
-                               if dec_successes_needed > 0 {
-                                       self.failed_channels.insert(chan_id, dec_successes_needed);
-                               } else {
-                                       self.failed_channels.remove(&chan_id);
-                               }
+                       let mut can_remove = false;
+                       if let Some(successes_needed) = self.failed_channels.get_mut(&chan_id) {
+                               successes_needed - 1;
+                               can_remove = if successes_neeed == 0 { true } else { false };
+                       }
+                       if can_remove {
+                               self.failed_channels.remove(&chan_id);
                        }
                }
        }
```

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -224,7 +224,7 @@ pub fn get_route<C: Deref, L: Deref>(our_node_id: &PublicKey, net_graph_msg_hand
 				let proportional_fee_millions = ($starting_fee_msat + final_value_msat).checked_mul($directional_info.fees.proportional_millionths as u64);
 				if let Some(new_fee) = proportional_fee_millions.and_then(|part| {
-						($directional_info.fees.base_msat as u64).checked_add(part / 1000000) })
+						($directional_info.fees.base_msat as u64).checked_add(net_metadata.get_channel_fee_penalty($chan_id.clone()) + part/1000000) })
```
Please used checked_add instead of + so that we refuse to use the channel on overflow.

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -428,6 +428,75 @@ impl Readable for NodeInfo {
+	fn route_failed(&mut self, route: Vec<RouteHop>, failed_hop: RouteHop) {
+		for route_hop in route {
+			if route_hop == failed_hop {
+				*self.failed_channels.entry(failed_hop.short_channel_id).or_insert(5) += 1;
```
TBH this feels a bit weird and arbitrary - you need 5 successes to un-fail a channel but once we have even 1 failure we refuse to route through a channel at all. I think maybe a more interesting approach would be to consider the time of the last failure and de-prioritize exponentially less as a failure ages, removing it entirely if we succeed. Because we cannot call current-time functions in the core library, we can't do that exactly, however something that may approximate would be to fail a channel until the next success - ie if we fail a channel, we wont use it the next time we route, but we'll use it any time after that. A user could, of course, do something better themselves by re-implementing the API. Separately, we could maybe add a feature flag for "we can call gettimeofday()".

sourabhmarathe replied:
Yes, I agree that it feels weird/arbitrary. I like the idea of having the exponential backoff too, but as mentioned we do not have access to time. For now, I am implementing "the opposite" where on success routing, all channels are removed from the failed_channels list, and on failure, channels have a "consecutive fails since last successful payment" count incremented.

TheBlueMatt replied:
Right, but because we fully-refuse to route through a channel after its failed once, we should never be able to send through it ever again if it fails once (because we'll never succeed routing through it if we never try...).

TheBlueMatt commented:
This needs rebase.

jkczyz commented:
Closing as this was mostly implemented across #1124, #1133, #1144, and #1146.
